In [1]:
from sklearn import datasets
from sklearn.linear_model import SGDRegressor
from sklearn.model_selection import train_test_split

In [2]:
import pandas as pd
import numpy as np

In [4]:
data = datasets.fetch_california_housing()
data

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
           37.88      , -122.23      ],
        [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
           37.86      , -122.22      ],
        [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
           37.85      , -122.24      ],
        ...,
        [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
           39.43      , -121.22      ],
        [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
           39.43      , -121.32      ],
        [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
           39.37      , -121.24      ]]),
 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]),
 'frame': None,
 'target_names': ['MedHouseVal'],
 'feature_names': ['MedInc',
  'HouseAge',
  'AveRooms',
  'AveBedrms',
  'Population',
  'AveOccup',
  'Latitude',
  'Longitude'],
 'DESCR': '.. _california_housing_dataset:\n

In [5]:
x = data.data
y = data.target
# columns_name = data.feature_names --- no need here

In [14]:
# applying standard scaler as SGD is scale sensitive

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
x = scaler.fit_transform(x)

In [15]:
# splitting the dataset

x_train, x_test, y_train, y_test = train_test_split(x,y)

In [16]:
print(x_train.shape)
print(x_test.shape)

(15480, 8)
(5160, 8)


In [17]:
# creating the object of sgd

algo = SGDRegressor()

In [18]:
# initial training using training data

algo.fit(x_train,y_train)

SGDRegressor()

In [19]:
print(algo.score(x_train,y_train))
print(algo.coef_)
print(algo.intercept_)

0.612144407058218
[ 0.856117    0.11385365 -0.32585818  0.37792113  0.0096277  -0.11877075
 -0.88373919 -0.87320026]
[2.05669683]


In [20]:
# lets again train the model by incrementing from now on with test data

algo.partial_fit(x_test,y_test)

SGDRegressor()

In [21]:
print(algo.score(x_train,y_train))
print(algo.coef_)
print(algo.intercept_)

0.5964027679100192
[ 0.83013354  0.12408628 -0.25060867  0.36095933  0.00669925 -0.36110906
 -0.88270651 -0.8615804 ]
[2.0551054]


Why Your Score Dropped (The Direct Answer)
Cell 18: algo.fit(x_train, y_train)

You trained your model on the x_train data.
The model worked hard to find the best possible parameters (coef_ and intercept_) to get a good score for this specific data.

Score on x_train: 0.612

Cell 20: algo.partial_fit(x_test, y_test)

You then told the model to update itself using the x_test data.
The patterns in x_test are not identical to the patterns in x_train.
So, the model changed its parameters to get better at predicting x_test. It "moved away" from what it learned from x_train.

Cell 21: print(algo.score(x_train, y_train))

You checked the score on x_train again.
Because the model's parameters have been changed to fit x_test, they are no longer the "perfect" parameters for x_train.
Score on x_train: 0.596 (It dropped, just as we'd expect!)

Analogy: You trained the model to be a x_train specialist. Then, you re-trained it to be a x_test specialist. This new specialty made it slightly worse at its original job.